src/subscribers.rs

```diff
@@ -0,0 +1,136 @@
+use packet::Blob;
+use rayon::prelude::*;
+use result::{Error, Result};
+use std::net::{SocketAddr, UdpSocket};
+
+#[derive(Clone, PartialEq)]
+pub struct Node {
+    pub id: [u64; 8],
+    pub weight: u64,
+    pub addr: SocketAddr,
+}
+
+//sockaddr doesn't implement default
+impl Default for Node {
+    fn default() -> Node {
+        Node {
+            id: [0; 8],
+            weight: 0,
+            addr: "0.0.0.0:0".parse().unwrap(),
+        }
+    }
+}
+
+impl Node {
+    pub fn new(id: [u64; 8], weight: u64, addr: SocketAddr) -> Node {
+        Node {
+            id: id,
+            weight: weight,
+            addr: addr,
+        }
+    }
+    fn key(&self) -> i64 {
+        (self.weight as i64).checked_neg().unwrap()
+    }
+}
+
+pub struct Subscribers {
+    pub data: Vec<Node>,
+    pub me: Node,
+    pub leader: Node,
+}
+
+impl Subscribers {
+    pub fn new(me: Node, leader: Node) -> Subscribers {
+        let mut h = Subscribers {
+            data: vec![],
+            me: me.clone(),
+            leader: leader.clone(),
+        };
+        h.insert(&[me, leader]);
```

Comment:
If you make this `&[&h.me, &h.leader]` can you remove the clones above()?

Reply:
its the most tiny amount of data

Reply:
```
error[E0502]: cannot borrow `h.me` as immutable because `h` is also borrowed as mutable
  --> src/subscribers.rs:54:21
   |
54 |         h.insert(&[&h.me, &h.leader]);
   |         -           ^^^^            - mutable borrow ends here             
   |         |           |                                                                          
   |         |           immutable borrow occurs here                                                                                                         
   |         mutable borrow occur
```
it generally doesn't like borrowing two fields from a structure

Reply:
this needs a copy, since we keep the node in the subscriber vector, and part of the structure